clients/roscpp/include/ros/timer_manager.h

```diff
@@ -428,11 +428,36 @@ void TimerManager<T, D, E>::setPeriod(int32_t handle, const D& period)
   {
     return;
   }
+  
+  if(reset)
+  {
+    info->next_expected = T::now() + period;
+  }
+  
+  // If called in a callback, last_real has not been updated yet so (now - last_real) > period
+  // In this case, let next_expected be updated only in updateNext
```

Comment:
Actually I don't think this check is needed at all.

* If a callback calls `setPeriod` of an unrelated timer it can just updated the `next_expected` value(either with `now` or `last_real + period`).

* If a callbacks calls `setPeriod` on its own timer The only thing that matter is the new period since the `next_expected` will be updated immediately after the invocation of the callback returns in the `updateNext` function.

Therefore I would think the following should work:

    {
      boost::mutex::scoped_lock lock(waiting_mutex_);
      info->period = period;

      if (reset)
      {
        // schedule the next invocation "period" from now
        info->next_expected = T::now() + period;
      }
      else
      {
        if (T::now() - info->last_real > period)
        {
          // the timer has already elapsed according to new period
          // schedule the next invocation immediately
          info->next_expected = T::now();
        }
        }
        else
        {
          // schedule the next invocation based on the last invocation + period
          info->next_expected = info->last_real + period;
        }
      }

      waiting_.sort(boost::bind(&TimerManager::waitingCompare, this, _1, _2));
    }